### ISSUE TITLE
Realtek 2.5G LAN USB Lan (r8156) dmesg/performance

### ISSUE THREAD

**Issue by joshuataylor** [Joshua-Riek/ubuntu-rockchip#278]
Howdy,

I have a 2.5GB LAN USB device ([USB to 2.5G Ethernet Adapter](https://www.cablematters.com/pc-1287-138-usb-to-25g-ethernet-adapter.aspx), it was cheap on Amazon as many of them are :)), which uses the [Realtek r8156 controller](https://www.realtek.com/en/component/zoo/category/network-interface-controllers-10-100-1000m-gigabit-ethernet-usb-3-0-software). 

This works out of the box with the Orange Pi 5 when plugged into USB3. However, dmesg is spammed with the following log when the interface is up, around every 500ms:

> cdc_ncm 6-1:2.0 enx5c857e343bf0: 2500 mbit/s downlink 2500 mbit/s uplink

I'm also seeing slightly higher latency when doing a simple ping test than other devices I've tested this on (AMD64 ArchLinux, Windows 10, MacOS), where the latency 1.5ms-2.5ms compared to ~0.5ms on other devices. Not sure if this is an Orange Pi 5 issue though. iPerf3 shows a sim

> Note: This is a USB A port, I have another dongle with just the USBC and it exhibits the same behaviour.

<details><summary>lsusb</summary>
<p>

```shell
Bus 006 Device 002: ID 0bda:8156 Realtek Semiconductor Corp. USB 10/100/1G/2.5G LAN
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               3.20
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         9
  idVendor           0x0bda Realtek Semiconductor Corp.
  idProduct          0x8156 
  bcdDevice           31.00
  iManufacturer           1 Realtek
  iProduct                2 USB 10/100/1G/2.5G LAN
  iSerial                 6 001000001
  bNumConfigurations      3
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0039
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              256mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           3
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass    255 Vendor Specific Subclass
      bInterfaceProtocol      0 
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval              11
        bMaxBurst               0
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0068
    bNumInterfaces          2
    bConfigurationValue     2
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              256mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass     13 
      bInterfaceProtocol      0 
      iInterface              5 CDC Communications Control
      CDC Header:
        bcdCDC               1.10
      CDC Union:
        bMasterInterface        0
        bSlaveInterface         1 
      CDC Ethernet:
        iMacAddress                      3 5C857E343BF0
        bmEthernetStatistics    0x0031501f
        wMaxSegmentSize               1518
        wNumberMCFilters            0x8000
        bNumberPowerFilters              0
      CDC NCM:
        bcdNcmVersion        1.00
        bmNetworkCapabilities 0x2b
          8-byte ntb input size
          max datagram size
          net address
          packet filter
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval              11
        bMaxBurst               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           0
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 
      iInterface              0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       1
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 
      iInterface              4 Ethernet Data
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0062
    bNumInterfaces          2
    bConfigurationValue     3
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              256mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         2 Communications
      bInterfaceSubClass      6 Ethernet Networking
      bInterfaceProtocol      0 
      iInterface              5 CDC Communications Control
      CDC Header:
        bcdCDC               1.10
      CDC Union:
        bMasterInterface        0
        bSlaveInterface         1 
      CDC Ethernet:
        iMacAddress                      3 5C857E343BF0
        bmEthernetStatistics    0x0031501f
        wMaxSegmentSize               1518
        wNumberMCFilters            0x8000
        bNumberPowerFilters              0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval              11
        bMaxBurst               0
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           0
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       1
      bNumEndpoints           2
      bInterfaceClass        10 CDC Data
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              4 Ethernet Data
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0400  1x 1024 bytes
        bInterval               0
        bMaxBurst               3
Binary Object Store Descriptor:
  bLength                 5
  bDescriptorType        15
  wTotalLength       0x0016
  bNumDeviceCaps          2
  USB 2.0 Extension Device Capability:
    bLength                 7
    bDescriptorType        16
    bDevCapabilityType      2
    bmAttributes   0x00000002
      HIRD Link Power Management (LPM) Supported
  SuperSpeed USB Device Capability:
    bLength                10
    bDescriptorType        16
    bDevCapabilityType      3
    bmAttributes         0x02
      Latency Tolerance Messages (LTM) Supported
    wSpeedsSupported   0x000e
      Device can operate at Full Speed (12Mbps)
      Device can operate at High Speed (480Mbps)
      Device can operate at SuperSpeed (5Gbps)
    bFunctionalitySupport   2
      Lowest fully-functional device speed is High Speed (480Mbps)
    bU1DevExitLat          10 micro seconds
    bU2DevExitLat        2047 micro seconds
Device Status:     0x0010
  (Bus Powered)
  Latency Tolerance Messaging (LTM) Enabled
```
</p>
</details>

<details>
<summary>sudo iperf3 -c 10.1.1.2 -t 30 -p 9000</summary>
```bash
[  5] local 10.1.0.138 port 40126 connected to 10.1.1.2 port 9000
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   144 MBytes  1.21 Gbits/sec    0   1.15 MBytes
[  5]   1.00-2.00   sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]   2.00-3.00   sec   139 MBytes  1.16 Gbits/sec    0   1.15 MBytes
[  5]   3.00-4.00   sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]   4.00-5.00   sec   139 MBytes  1.16 Gbits/sec    0   1.15 MBytes
[  5]   5.00-6.00   sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]   6.00-7.00   sec   139 MBytes  1.16 Gbits/sec    0   1.15 MBytes
[  5]   7.00-8.00   sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]   8.00-9.00   sec   139 MBytes  1.16 Gbits/sec    0   1.15 MBytes
[  5]   9.00-10.00  sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]  10.00-11.00  sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]  11.00-12.00  sec   138 MBytes  1.15 Gbits/sec    0   1.15 MBytes
[  5]  12.00-13.00  sec   138 MBytes  1.15 Gbits/sec    0   1.21 MBytes
[  5]  13.00-14.00  sec   135 MBytes  1.13 Gbits/sec    0   1.21 MBytes
[  5]  14.00-15.00  sec   135 MBytes  1.13 Gbits/sec    0   1.21 MBytes
[  5]  15.00-16.00  sec   148 MBytes  1.24 Gbits/sec    0   1.21 MBytes
[  5]  16.00-17.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  17.00-18.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  18.00-19.00  sec   148 MBytes  1.24 Gbits/sec    0   1.21 MBytes
[  5]  19.00-20.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  20.00-21.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  21.00-22.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  22.00-23.00  sec   148 MBytes  1.24 Gbits/sec    0   1.21 MBytes
[  5]  23.00-24.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  24.00-25.00  sec   148 MBytes  1.24 Gbits/sec    0   1.21 MBytes
[  5]  25.00-26.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  26.00-27.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  27.00-28.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  28.00-29.00  sec   146 MBytes  1.23 Gbits/sec    0   1.21 MBytes
[  5]  29.00-30.00  sec   148 MBytes  1.24 Gbits/sec    0   1.21 MBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-30.00  sec  4.17 GBytes  1.19 Gbits/sec    0             sender
[  5]   0.00-30.00  sec  4.17 GBytes  1.19 Gbits/sec                  receiver
```
</details>

These Realtek devices can be pretty hit and miss, but for the price they're great value.

**Comment by joshuataylor** [comment]
Fixed by using the drivers from Realtek, similar to how this fixes many issues with their PCI cards :-). Didn't realise it was the same driver.

tldr:
```sh
sudo apt update
sudo apt upgrade
```
# REBOOT IF YOU INSTALLED A NEW KERNEL

```sh
sudo apt install build-essential dkms
git clone https://github.com/awesometic/realtek-r8125-dkms --depth=1
cd realtek-r8125-dkms
sudo ./dkms-install.sh
```

This will now autoupdate on kernel updates 🎉.
